doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,221 @@
+| **streaming request** | client-streaming   | bidirectional (bidi)   |
+
+While we use the term "method," pedantically it is closer to "function" as it is
+not object oriented and there is no "receiver" involved (the `this` variable in
+many languages) other than the destination machine. gRPC is based on message
+passing, not object orientation.
+
+Related Methods are typically grouped into a Service. To gRPC, a Service is a
+group of methods that tend to be implemented together and that all share the
+Service's namespace. A Service is a higher-level abstraction and may not be
+present explicitly in all implementations. However, the namespace provided by a
+Service is a core distinguishing feature of its Methods; if two Methods have the
+same name but exist in different Services they must be considered distinct and
+not be confused. A Method name including its Service namespace prefix with a "/"
+separator is a "full method name".
+
+## Calls
+
+RPCs, or "Calls," are initiated by a client to a server, typically via a
+Channel. The initiation is with Request Headers, within which the client
+indicates the method to be run by its Full Method Name. The Call is gracefully
+completed when the server responds with Trailers, which contains a Status
+communicating the success or failure of the RPC. Note that on the server there
+is a period of time between when the server application responds with a Trailers
+and when that Trailers is actually sent; the Call is only truly complete when
+the Trailers is sent. Similarly, on the client there is a period between the
+gRPC implementation receiving the Trailers and when the application receives the
+Trailers; the Call is only truly complete when the Trailers is received by the
+application.
+
+Calls may terminate early by being "cancelled." Implementations must allow
+clients to cancel Calls, but cancellations may occur in other ways like I/O
+failures. A cancellation appears as a Trailers to clients and is a special state
+on servers. Cancellation is an abrupt killing of the Call; inbound and outbound
+buffered data should be cleared. Cancellation trumps graceful completion; if the
+client gRPC implementation received the Trailers before the cancellation, yet
+the client application has not received the Trailers, then cancellation
+generally should win. No auxilary information is included in cancellation
+signals between the client and server. Server implementations may fail a Call
+and respond with Trailers while claiming to the server application that the Call
+was cancelled.
+
+There may be multiple servers that _could_ have received the Call (as is common
+for load balancing), but only a single server may process an individual Call.
+Calls are assumed to be non-idempotent and may not be "replayed" except for when
+gRPC is explicitly informed it is safe to do so.
+
+Calls are natively two independent streams (i.e., full duplex bidirectional) of
```

Comment:
This paragraph is great!  It's a really clear description of the overall structure of a call.  I think we should actually move this up to be the second paragraph of this section, right after the one about how calls map from client to server.

Reply:
Moved and reworked this area.